### PR TITLE
Refactor/separate block production and running

### PR DESCRIPTION
**Pull request by asdacap**
- Separate the building of block and the triggering of the building of the block.
- Make it easy to identify if a plugin is replacing only the building of the block or signalling of the block production, which does not change a lot actually.
- Remove the need of IManualBlockProductionTrigger and just call IBlockProducer directly, except of `evm_mine` which may want to trigger event which is used to put the block in main chain.

## Changes

- Created IBlockProducerRunner which have the interface of the current IBlockProducer.
- The interface of IBlockProducer is now the same as IManualBlockProductionTrigger.
- Move any use of IManualBlockProductionTrigger to just calling IBlockProducer directly.

#### What types of changes does your code introduce?

- [X] Refactoring

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [X] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Consensus/IBlockProducer.cs [hunk 1]
```diff
@@ -1,16 +1,19 @@
...
 
 public interface IBlockProducer
 {
-    Task Start();
-    Task StopAsync();
-    bool IsProducingBlocks(ulong? maxProducingInterval);
-    event EventHandler<BlockEventArgs> BlockProduced;
+    Task<Block?> BuildBlock(
+        BlockHeader? parentHeader = null,
+        CancellationToken? cancellationToken = null,
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/IBlockProducer.cs:16]: minor: convention puts cancellationToken often as last parameter to separate it from more logic-oriented parameters.

File: src/Nethermind/Nethermind.Consensus/IBlockProducerRunner.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+using System;
+using System.Threading.Tasks;
+using Nethermind.Core;
+
+namespace Nethermind.Consensus;
+
+public interface IBlockProducerRunner
+{
+    Task Start();
+    Task StopAsync();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/IBlockProducerRunner.cs:13]: We don't have truly async implementations of this?

File: src/Nethermind/Nethermind.Consensus/Producers/IBlockProducerInfo.cs [hunk 3]
```diff
@@ -1,14 +1,16 @@
...
+using Nethermind.Core;
 using Nethermind.Evm.Tracing;
 
 namespace Nethermind.Consensus.Producers
 {
     public interface IBlockProducerInfo
     {
         IBlockProducer BlockProducer { get; }
-        IManualBlockProductionTrigger BlockProductionTrigger { get; }
+        Func<BlockHeader, bool> Condition { get; }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Producers/IBlockProducerInfo.cs:13]: Cryptic?

**Review by LukaszRozmej — COMMENTED** [review]
Mixed feelings: It doesn't seem to make things simpler? Also am I right that we are loosing easily triggering block production from a simple plugin as we are not exposing trigger?

**Comment by asdacap** [comment]
Yea, it does not look like it help much unfortunately. It does expose `IConsensusPlugin.CreateBlockProducerRunner` for the purpose of the plugin orchestrating when to produce block. IMO it make it easier for the plugin to arbitrarily create block without creating side effect by calling `IBlockProducer.BuildBlock`.

File: src/Nethermind/Nethermind.Consensus/IBlockProducerRunner.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Consensus/IBlockProducerRunner.cs:13]: Only clique for stop

File: src/Nethermind/Nethermind.Consensus/Producers/IBlockProducerInfo.cs (hunk 3 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Consensus/Producers/IBlockProducerInfo.cs]: Switched to custom interface.

File: src/Nethermind/Nethermind.Consensus/IBlockProducerRunner.cs [hunk 4]
```diff
@@ -9,7 +9,7 @@ namespace Nethermind.Consensus;
 
 public interface IBlockProducerRunner
 {
-    Task Start();
+    void Start();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/IBlockProducerRunner.cs:12]: or we could make it a ValueTask if we wanted flexibility

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**